field-value-counter-sink/src/main/java/org/springframework/cloud/stream/module/metrics/FieldValueCounterSink.java

```diff
@@ -0,0 +1,141 @@
+ */
+@EnableBinding(Sink.class)
+public class FieldValueCounterSink {
+
+	@Autowired
+	private FieldValueCounterSinkProperties fvcSinkProperties;
+
+	@Autowired
+	private FieldValueCounterRepository fieldValueCounterRepository;
+
+	@Autowired
+	@Qualifier(IntegrationContextUtils.INTEGRATION_EVALUATION_CONTEXT_BEAN_NAME)
+	private EvaluationContext evaluationContext;
+
+	private final JsonToTupleTransformer jsonToTupleTransformer = new JsonToTupleTransformer();
+
+	@ServiceActivator(inputChannel=Sink.INPUT)
+	public void process(Message<?> message) {
+		Object payload = message.getPayload();
+		if (payload instanceof String) {
+			try {
+				payload = jsonToTupleTransformer.transformPayload(payload.toString());
+			}
+			catch (Exception e) {
+				throw new MessageTransformationException(message, e.getMessage(), e);
+			}
+		}
+		if (payload instanceof Tuple) {
+			processTuple(computeMetricName(message), (Tuple) payload);
+		}
+		else {
+			processPojo(computeMetricName(message), payload);
+		}
+	}
+
+	private void processPojo(String counterName, Object payload) {
+		BeanWrapper beanWrapper = new BeanWrapperImpl(payload);
+		if (beanWrapper.isReadableProperty(fvcSinkProperties.getFieldName())) {
+			Object value = beanWrapper.getPropertyValue(fvcSinkProperties.getFieldName());
+			processValue(counterName, value);
+		}
+	}
+
+	private void processTuple(String counterName, Tuple tuple) {
+		String[] path = StringUtils.tokenizeToStringArray(fvcSinkProperties.getFieldName(), ".");
+		processValueForCounter(counterName, tuple, path);
+	}
+
+	private void processValueForCounter(String counterName, Object value, String[] path) {
+		String key = path[0];
+		Object result = null;
+		if (value instanceof List) {
+			for (Object item : (List<?>) value) {
+				processValueForCounter(counterName, item, path);
+			}
+		}
+		else if (value instanceof Tuple) {
+			Tuple t = (Tuple) value;
+			if (t.hasFieldName(key)) {
+				result = t.getValue(key);
+			}
+		}
+		else if (value instanceof Map) {
+			result = ((Map<?, ?>) value).get(key);
+		}
+		if (result != null) {
+			if (path.length == 1) {
+				processValue(counterName, result);
+			}
+			else {
+				path = Arrays.copyOfRange(path, 1, path.length);
+				processValueForCounter(counterName, result, path);
+			}
+		}
+	}
+
+	protected void processValue(String counterName, Object value) {
+		if ((value instanceof Collection) || ObjectUtils.isArray(value)) {
+			Collection<?> c = (value instanceof Collection) ? (Collection<?>) value
+					: Arrays.asList(ObjectUtils.toObjectArray(value));
+			for (Object val : c) {
+				// TODO better conversion to a string
+				fieldValueCounterRepository.increment(counterName, val.toString(), 1.0);
+			}
+		}
+		else {
+			fieldValueCounterRepository.increment(counterName, value.toString(), 1.0);
+		}
+	}
+
+	protected String computeMetricName(Message<?> message) {
+		return fvcSinkProperties.getComputedNameExpression().getValue(evaluationContext,
```

Comment:
EvaluationContext should already be set thanks to `SpelExpressionConverterConfiguration`